extensions/standard-processors/tests/unit/PutTCPTests.cpp

```diff
@@ -0,0 +1,454 @@
+
+  receive_success(test_fixture, first_message);
+  receive_success(test_fixture, second_message);
+  receive_success(test_fixture, third_message);
+
+  CHECK(1 == test_fixture.getNumberOfActiveSessions());
+
+  test_fixture.closeActiveConnections();
+
+  trigger_expect_success(test_fixture, fourth_message);
+  trigger_expect_success(test_fixture, fifth_message);
+  trigger_expect_success(test_fixture, sixth_message);
+
+  test_fixture.tryDequeueReceivedMessage();
+
+  CHECK(LogTestController::getInstance().matchesRegex("warning.*with reused connection, retrying"));
+  CHECK(2 == test_fixture.getNumberOfActiveSessions());
+}
+
+void test_connection_per_flow_file(PutTCPTestFixture& test_fixture) {
+  test_fixture.enableConnectionPerFlowFile();
+
+  trigger_expect_success(test_fixture, first_message);
+  trigger_expect_success(test_fixture, second_message);
+  trigger_expect_success(test_fixture, third_message);
+
+  receive_success(test_fixture, first_message);
+  receive_success(test_fixture, second_message);
+  receive_success(test_fixture, third_message);
+
+  trigger_expect_success(test_fixture, fourth_message);
+  trigger_expect_success(test_fixture, fifth_message);
+  trigger_expect_success(test_fixture, sixth_message);
+
+  receive_success(test_fixture, fourth_message);
+  receive_success(test_fixture, fifth_message);
+  receive_success(test_fixture, sixth_message);
+
+  CHECK(6 == test_fixture.getNumberOfActiveSessions());
+}
+
+void test_idle_expiration(PutTCPTestFixture& test_fixture) {
+  test_fixture.setIdleConnectionExpiration("100ms");
+  trigger_expect_success(test_fixture, first_message);
+  std::this_thread::sleep_for(110ms);
+  trigger_expect_success(test_fixture, second_message);
+
+  receive_success(test_fixture, first_message);
+  receive_success(test_fixture, second_message);
+
+  CHECK(2 == test_fixture.getNumberOfActiveSessions());
+}
+
+TEST_CASE("Server closes in-use socket", "[PutTCP]") {
+  PutTCPTestFixture test_fixture;
+  SECTION("No SSL") {
+    test_fixture.startTCPServer();
+    test_dropped_in_use_socket(test_fixture);
+  }
+  SECTION("SSL") {
+    test_fixture.addSSLContextToPutTCP("ca_A.crt", "alice_by_A.pem");
+    test_fixture.startSSLServer();
+
+    test_dropped_in_use_socket(test_fixture);
+  }
+}
+
+TEST_CASE("Connection per flow file", "[PutTCP]") {
+  PutTCPTestFixture test_fixture;
+  SECTION("No SSL") {
+    test_fixture.startTCPServer();
+    test_connection_per_flow_file(test_fixture);
+  }
+  SECTION("SSL") {
+    test_fixture.addSSLContextToPutTCP("ca_A.crt", "alice_by_A.pem");
+    test_fixture.startSSLServer();
+
+    test_connection_per_flow_file(test_fixture);
+  }
+}
+
+TEST_CASE("PutTCP test invalid host", "[PutTCP]") {
+  PutTCPTestFixture test_fixture;
+  SECTION("No SSL") {
+    test_invalid_host(test_fixture);
+  }
+  SECTION("SSL") {
+    test_fixture.addSSLContextToPutTCP("ca_A.crt", "alice_by_A.pem");
+    test_invalid_host(test_fixture);
```

Comment:
the functions `text_*` seem to be single use, we could move the `test_invalid_host` calls (and other single-use functions) out of the sections and put the implementation after the sections, I believe they would be equivalent

Reply:
makes sense, I've changed these in https://github.com/apache/nifi-minifi-cpp/pull/1419/commits/cf6e53eecf233d1d38553e81e71a70aaad2738e7#diff-bbb56b112b398a6450c23e71e5dbbfb17bf20dad3d2fb908a1c6de26118f9ee8L256-L301